File: knigs_of_the_forest.cpp
/*
Problem: https://open.kattis.com/problems/knigsoftheforest
By: Jens Ekenblad
Date: 2023-06-14
*/

#include <iostream>
#include <queue>
#include <unordered_map>
#include <vector>
#include <string>
#include <sstream>

using namespace std;

int main()
{
    int k, n;
    cin >> k >> n;
    int y, p;
    cin >> y >> p;

    unordered_map<int, priority_queue<int>> yearGraph;
    yearGraph[y].push(p);

    // Add each new contestant to the graph with its power
    string line;
    getline(cin, line);
    while (getline(cin, line))
    {
        istringstream iss(line);
        int mooseYear, moosePwr;
        iss >> mooseYear >> moosePwr;
        yearGraph[mooseYear].push(moosePwr);
    }

    // Iterate each year, remove the strongest moose than check if Karl-Älgtav is the strongest
    for (int year = 2011; year < 2011 + n; year++)
    {
        if (year != 2011)
        {
            yearGraph[2011].push(yearGraph[year].top());
        }
        int max_power = yearGraph[2011].top();
        yearGraph[2011].pop();
        if (max_power == p)
        {
            cout << year << endl;
            return 0;
        }
    }
    cout << "unknown" << endl;
}

File: knigs_of_the_forest.py
# Problem: https://open.kattis.com/problems/knigsoftheforest
# By: Jens Ekenblad
# Date: 2023-06-14

import sys
import heapq

lines=sys.stdin.readlines()
k, n = map(int, lines[0].split())   # k = number of moose, n = number of years
y, p = map(int, lines[1].split())   # y = Karl-Älgtav year, p = Karl-Älgtav power

yearGraph = {}
yearGraph[y] = [-p]

# Add each new contestant to the graph with its power
for line in lines[2:]:
    mooseYear, moosePwr = map(int, line.split())
    if mooseYear not in yearGraph:
        yearGraph[mooseYear] = [-moosePwr]
    else:
        yearGraph[mooseYear].append(-moosePwr)


heapq.heapify(yearGraph[2011])
# Iterate each year, remove the strongest moose than check if Karl-Älgtav is the strongest
for year in range(2011, 2011+n):
    if year != 2011:
        heapq.heappush(yearGraph[2011], heapq.heappop(yearGraph[year]))
    
    max_power = -heapq.heappop(yearGraph[2011])
    if max_power == p:
        print(year)
        quit()
        
print("unknown")
    